src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -240,13 +240,17 @@ module.exports = function<T, P, I, TI, C, CX>(
     }
 
     // Rerender
-    const instance = workInProgress.stateNode;
-    ReactCurrentOwner.current = workInProgress;
-    const nextChildren = instance.render();
-    reconcileChildren(current, workInProgress, nextChildren);
-    // Put context on the stack because we will work on children
-    if (isContextProvider(workInProgress)) {
-      pushContextProvider(workInProgress, true);
+    try {
+      const instance = workInProgress.stateNode;
+      ReactCurrentOwner.current = workInProgress;
+      const nextChildren = instance.render();
+      reconcileChildren(current, workInProgress, nextChildren);
+    } finally {
```

Comment:
I could move the call to `instance.render` into a separate function that is try/finally wrapped to minimize the impact. (I'm not an expert with the whole inlining area.)

I'm also working today on an effort to combine `ReactFiberContext` and `ReactFiberHostContext` and add fiber-level tracking so that we won't pop a Fiber more than once, or pop one that wasn't pushed. It will hopefully ease the pressure here a bit as it would enable us to just dev-warn and ignore the bad pop rather than throw an error.